src/dh/dhe.ts

```diff
@@ -0,0 +1,235 @@
+ */
+export async function createInteractiveConsoleQuery(
+  tagId: UniqueID,
+  dheClient: EnterpriseClient,
+  consoleType?: ConsoleType
+): Promise<QuerySerial> {
+  const userInfo = await dheClient.getUserInfo();
+  const owner = userInfo.username;
+  const type = INTERACTIVE_CONSOLE_QUERY_TYPE;
+
+  const [dbServers, queryConstants, serverConfigValues] = await Promise.all([
+    dheClient.getDbServers(),
+    dheClient.getQueryConstants(),
+    dheClient.getServerConfigValues(),
+  ]);
+
+  const name = `IC - VS Code - ${tagId}`;
+  const dbServerName = dbServers[0]?.name ?? 'Query 1';
+  const heapSize = queryConstants.pqDefaultHeap;
+  const jvmProfile = serverConfigValues.jvmProfileDefault;
+  const scriptLanguage =
+    serverConfigValues.scriptSessionProviders?.find(
+      p => p.toLocaleLowerCase() === consoleType
+    ) ?? 'Python';
+  const workerKind = serverConfigValues.workerKinds?.[0]?.name;
+
+  const timeZone =
+    serverConfigValues.timeZone ??
+    Intl.DateTimeFormat().resolvedOptions().timeZone;
+
+  const scheduling = QueryScheduler.makeDefaultScheduling(
+    serverConfigValues.restartQueryWhenRunningDefault,
+    timeZone
+  );
+
+  const draftQuery = new DraftQuery({
+    isClientSide: true,
+    draftOwner: owner,
+    name,
+    type,
+    owner,
+    dbServerName,
+    heapSize: heapSize,
+    scheduling,
+    // TODO: deephaven/vscode-deephaven/issues/153 to update this to secure websocket connection
+    jvmArgs: '-Dhttp.websockets=true',
+    jvmProfile,
+    scriptLanguage,
+    workerKind,
+  });
+
+  if (draftQuery.serial == null) {
+    draftQuery.updateSchedule();
+  }
+
+  // type assertion gives us stronger type safety than the Promise<string>
+  // return type defined by the JS API types.
+  return dheClient.createQuery(draftQuery) as Promise<QuerySerial>;
+}
+
+/**
+ * Delete queries by serial.
+ * @param dheClient DHE client to use.
+ * @param querySerials Serials of queries to delete.
+ */
+export async function deleteQueries(
+  dheClient: EnterpriseClient,
+  querySerials: QuerySerial[]
+): Promise<void> {
+  await dheClient.deleteQueries(querySerials);
+}
+
+/**
+ * Get worker info from a query serial.
+ * @param tagId Unique tag id to include in the worker info.
+ * @param dhe DHE JsApi instance.
+ * @param dheClient DHE client to use.
+ * @param querySerial Serial of the query to get worker info for.
+ * @returns A promise that resolves to the worker info when the worker is ready.
+ */
+export async function getWorkerInfoFromQuery(
+  tagId: UniqueID,
+  dhe: DheType,
+  dheClient: EnterpriseClient,
+  querySerial: QuerySerial
+): Promise<WorkerInfo | undefined> {
+  // The query will go through multiple config updates before the worker is ready.
+  // This Promise will respond to config update events and resolve when the worker
+  // is ready.
+  const queryInfo = await new Promise<QueryInfo>(resolve => {
+    dheClient.addEventListener(dhe.Client.EVENT_CONFIG_UPDATED, _event => {
```

Comment:
`EVENT_CONFIG_UPDATED` is fired fairly often and scanning known configs can be problematic on large query sets. I'd do an initial `findQueryConfigForSerial` and subscribe to the update event if the info object with the URL isn't found.

Reply:
Added to [Connection / event handler cleanup #152](https://github.com/deephaven/vscode-deephaven/issues/152)